### PR TITLE
Hash

### PR DESCRIPTION
**Pull request by WangBoxue**
Implement functions `keepIf`, `removeIf`.
`keepIf` and `filter` are similar. `removeIf` and `reject` are similar.
The difference is that `keepIf` and `removeIf` take effect on the hash itself, while `filter` and `reject` return a new hash.

### CODE REVIEW

**Comment by LeonLiuY** [comment]
Maybe we can also choose to use `reject$` `filter$`. 

How do you think?

Since in `Seq`, we use this way

**Comment by WangBoxue** [comment]
@liuyang1204 done

File: src/test/java/com/worksap/fig/HashTest.java [hunk 1]
```diff
@@ -59,4 +59,31 @@ public void testFilter(){
...
+        hash.put(2, 4);
+        hash.put(5, 1);
+        assertEquals(2, hash.filter$((k, v) -> !(k > 1 && v > 1)).size());
+        assertEquals(null, hash.get(2));
+        assertEquals(2, hash.filter$((k, v) -> k != v).size());
+        assertEquals(Hash.newHash(), hash.filter$((k, v) -> k == v));
+        hash.filter$(null);
+    }
+
 }
```
Comment (LeonLiuY) [inline src/test/java/com/worksap/fig/HashTest.java:103]: Could you modify test case for `testReject` `testFilter` (not the two you committed), to ensure they don't change the original object?

Comment (WangBoxue) [inline src/test/java/com/worksap/fig/HashTest.java:103]: done

**Pull request merged**